codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,769 @@
+            throw new NumberFormatException();
+        }
+
+        int i = start;
+        boolean negative = byteAt(i) == '-';
+        if (negative && ++i == end) {
+            throw new NumberFormatException(subSequence(start, end).toString());
+        }
+
+        return parseLong(i, end, radix, negative);
+    }
+
+    private long parseLong(int start, int end, int radix, boolean negative) {
+        final byte[] value = this.value;
+        long max = Long.MIN_VALUE / radix;
+        long result = 0;
+        int offset = start;
+        while (offset < end) {
+            int digit = Character.digit((char) (value[offset++] & 0xFF), radix);
+            if (digit == -1) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            if (max > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            long next = result * radix - digit;
+            if (next > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            result = next;
+        }
+        if (!negative) {
+            result = -result;
+            if (result < 0) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+        }
+        return result;
+    }
+
+    public char parseChar() {
+        return parseChar(0, length());
+    }
+
+    public char parseChar(int start, int end) {
+        return toString(start, end).charAt(0);
+    }
+
+    public short parseShort() {
+        return parseShort(0, length(), 10);
+    }
+
+    public short parseShort(int radix) {
+        return parseShort(0, length(), radix);
+    }
+
+    public short parseShort(int start, int end) {
+        return parseShort(start, end, 10);
+    }
+
+    public short parseShort(int start, int end, int radix) {
+        int intValue = parseInt(start, end, radix);
+        short result = (short) intValue;
+        if (result != intValue) {
+            throw new NumberFormatException(subSequence(start, end).toString());
+        }
+        return result;
+    }
+
+    public float parseFloat() {
+        return parseFloat(0, length());
+    }
+
+    public float parseFloat(int start, int end) {
+        return Float.parseFloat(toString(start, end));
+    }
+
+    public double parseDouble() {
+        return parseDouble(0, length());
+    }
+
+    public double parseDouble(int start, int end) {
+        return Double.parseDouble(toString(start, end));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof ByteString)) {
+            return false;
+        }
+
+        if (this == obj) {
+            return true;
+        }
+
+        ByteString that = (ByteString) obj;
+        if (length() != that.length() || hashCode() != that.hashCode()) {
+            return false;
+        }
+
+        byte[] thisValue = value;
```

Comment:
We could...but we are taking advantage of the cached hash code to short circuit the equality check.